### DISCUSSION TITLE
Get number of dimensions for an array

### DISCUSSION THREAD

**Discussion by cristianbuse** [Theadd/ds3x#27] [General]
Hi @Theadd ,

Consider replacing ```Is1dArray``` and ```Is2dArray``` with something like ```GetArrayDimsCount(arr) = 1``` and ```GetArrayDimsCount(arr) = 2``` where ```GetArrayDimsCount``` looks like this:  
```VBA
Private Function GetArrayDimsCount(ByRef arr As Variant) As Long
    Const MAX_DIMENSION As Long = 60 'VB limit
    Dim dimension As Long
    Dim tempBound As Long
    '
    On Error GoTo FinalDimension
    For dimension = 1 To MAX_DIMENSION
        tempBound = LBound(arr, dimension)
    Next dimension
FinalDimension:
    GetArrayDimsCount = dimension - 1
End Function
```

**Comment by Theadd** [comment]
Hi again @cristianbuse ,

As lately I've been experimenting with all these memory/safearray stuff since I found out about your VBA-LibMemory project, I just applied these methods to see if they could successfully be used to get the same results of your `GetArrayDimsCount` code from above, which they do, and with even better performance.

```vb
Public Function GetArrayDims(ByRef TargetArray As Variant) As Integer
    GetArrayDims = MemInt(ArrPtr(TargetArray))
End Function
```

This one is even faster but it is tied to a single type of array.

```vb
Public Function GetArrayDimsNotNot(ByRef TargetArray() As Long) As Integer
    GetArrayDimsNotNot = MemInt(Not Not TargetArray)
End Function
```

These are the perf-test results:
```py
# 1M Iterations
GetArrayDims: 0.57812s
GetArrayDimsNotNot: 0.14062s
GetArrayDimsCount: 1.00781s

# 10M Iterations
GetArrayDims: 5.79688s
GetArrayDimsNotNot: 1.46094s
GetArrayDimsCount: 10.58594s
```

I thought that it'd have a bigger negative impact in a single call on first run after restarting the app, due to initialization of remote memory and safe array structs, but it was negligible, less than `0.00004s`, where `GetArrayDimsCount` was around `0.000025s`. Yours is a nice one trick indeed.

I [might](https://github.com/Theadd/ds3x/blob/main/src/ds3x.accdb.src/modules/ArrayList.cls#L488-L490) have found a problem using your `MemCopy` while copying very large regions of memory but don't mind me yet, that code is still young and probably it was due to something else.

Thanks! =)

**Comment by cristianbuse** [comment]
Hi @Theadd ,

You probably won't be surprised to know that I've tried all these across the years and I decided to go with the ```GetArrayDimsCount``` method as in [here](https://github.com/cristianbuse/VBA-ArrayTools/blob/80c94b05fa760d25459c8922a9f0e9f56b79e833/src/LibArrayTools.bas#L838-L854) in my Array Tools library. I will explain why.

Let's start with ```Not Not TargetArray```. This basically returns the pointer to the underlying [```SAFEARRAY```](https://learn.microsoft.com/en-us/windows/win32/api/oaidl/ns-oaidl-safearray) structure as you already know - same as ```ArrPtr```. There are 2 big reasons why this is a big NO NO for me:
1) Relying on a bug is dangerous and leads to unpredictable behaviour. If Microsoft decides (probably not 😄) to fix the bug in a future VBA release, then code relying on the bug stops working and may leads to unwanted results which are worse than a crash
2) You cannot have just one method to handle all the array types, like you already mentioned

Because of the above 2 issues, ```Not Not``` is not something I would ever recommend using, even with performance gains, when exposing the method in a library like I did in ```LibArrayTools```.

There is also a catch. It causes a crash for uninitialized arrays if not careful as in the following code:
```VBA
Option Explicit

Sub TestNotNot()
    Dim arr1() As Long
    Dim arr2() As Long
    
    ReDim arr1(0)
    Debug.Print GetArrayDimsNotNot(arr1) 'Prints 1
    Debug.Print GetArrayDimsNotNot(arr2) 'Crash
End Sub

Public Function GetArrayDimsNotNot(ByRef TargetArray() As Long) As Long
    GetArrayDimsNotNot = MemInt(Not Not TargetArray)
End Function
```
However, this is easily fixed with a slight change:
```VBA
Option Explicit

Sub TestNotNotImproved()
    Dim arr1() As Long
    Dim arr2() As Long
    
    ReDim arr1(0)
    Debug.Print GetArrayDimsNotNot(arr1) 'Prints 1
    Debug.Print GetArrayDimsNotNot(arr2) 'Prints 0
End Sub

Public Function GetArrayDimsNotNot(ByRef TargetArray() As Long) As Long
    Dim ptr As LongPtr: ptr = CLngPtr(Not Not TargetArray)
    If ptr = 0 Then Exit Function
    GetArrayDimsNotNot = MemInt(ptr)
End Function
```

How about ```MemInt(ArrPtr(TargetArray))```?. Well, ```ArrPtr(TargetArray)``` is safe but we run into the same issue with uninitialized arrays so at the very least, the function should look like:
```VBA
Public Function GetArrayDims(ByRef TargetArray As Variant) As Long
    Dim ptr As LongPtr: ptr = ArrPtr(TargetArray)
    If ptr = 0 Then Exit Function
    GetArrayDims = MemInt(ptr)
End Function
```

You probably noticed, that I changed the return type for both functions from ```Integer``` to ```Long```. While ```Integer``` is useful when working with specific memory bytes and expecially custom data types (for byte alignment in memory), there is no need to use them in regular code. Read the **BUT** section of [this SO answer](https://stackoverflow.com/a/26409520/8488913).

***

So, ```GetArrayDimsNotNot``` is out of the question for me. However, I suppose this could work as a ```Private``` method inside your class as long as you are 'fine' with using a bug as a feature.

Many years ago, when I tested the perfomance of ```GetArrayDimsCount``` against reading the 2 bytes from the ```SAFEARRAY```, the ```GetArrayDimsCount``` was considerably faster than the latter.

However, since you obtained the exact opposite, I re-tested today on 2 Windows machines, one x32 and one x64 using this:
```VBA
Option Explicit

Sub TestSpeed()
    Dim arr1() As Long: ReDim arr1(0)
    Dim arr2() As Long
    Dim arr3(0, 0, 0, 0, 0) As Long
    Dim r As Long
    Dim i As Long
    Dim t As Double
    Const iterations As Long = 1000000
    '
    t = Timer
    For i = 1 To iterations
        r = GetArrayDims(arr1)
        r = GetArrayDims(arr2)
        r = GetArrayDims(arr3)
    Next i
    Debug.Print "GetArrayDims: " & Round(Timer - t, 3)
    '
    t = Timer
    For i = 1 To iterations
        r = GetArrayDimsNotNot(arr1)
        r = GetArrayDimsNotNot(arr2)
        r = GetArrayDimsNotNot(arr3)
    Next i
    Debug.Print "GetArrayDimsNotNot: " & Round(Timer - t, 3)
    '
    t = Timer
    For i = 1 To iterations
        r = GetArrayDimsCount(arr1)
        r = GetArrayDimsCount(arr2)
        r = GetArrayDimsCount(arr3)
    Next i
    Debug.Print "GetArrayDimsCount: " & Round(Timer - t, 3)
End Sub

Public Function GetArrayDims(ByRef TargetArray As Variant) As Long
    Dim ptr As LongPtr: ptr = ArrPtr(TargetArray)
    If ptr = 0 Then Exit Function
    GetArrayDims = MemInt(ptr)
End Function

Public Function GetArrayDimsNotNot(ByRef TargetArray() As Long) As Long
    Dim ptr As LongPtr: ptr = CLngPtr(Not Not TargetArray)
    If ptr = 0 Then Exit Function
    GetArrayDimsNotNot = MemInt(ptr)
End Function

Private Function GetArrayDimsCount(ByRef arr As Variant) As Long
    Const MAX_DIMENSION As Long = 60 'VB limit
    Dim dimension As Long
    Dim tempBound As Long
    '
    On Error GoTo FinalDimension
    For dimension = 1 To MAX_DIMENSION
        tempBound = LBound(arr, dimension)
    Next dimension
FinalDimension:
    GetArrayDimsCount = dimension - 1
End Function
```
and got the following results:

|Method (1M iterations) |x32     |x64    |
|------------------------- |-------|-------|
|GetArrayDims               |2.711s|2.398s|
|GetArrayDimsNotNot   |0.535s|0.492s|
|GetArrayDimsCount      |1.207s|3.934s|

As you can see ```GetArrayDimsCount``` beats ```GetArrayDims``` on a x32 VBA and I think when I tested this many years ago, I did not have x64. Obviously you were correct and ```GetArrayDims``` is faster on x64.

I was mentioning on some other discussion that I plan to rewrite the Array Tools library to take advantage of memory copying and I will probably use ```GetArrayDims``` on x64 only while on x32 will keep the original ```GetArrayDimsCount```.

Many thanks for the head up!

***

>I [might](https://github.com/Theadd/ds3x/blob/main/src/ds3x.accdb.src/modules/ArrayList.cls#L488-L490) have found a problem using your MemCopy while copying very large regions of memory but don't mind me yet, that code is still young and probably it was due to something else.

I designed ```MemCopy``` to achieve the same results as ```CopyMemory``` API in that it deals with overlapping memory just fine and it's faster for VBA7 (```CopyMemory``` is still faster for VBA6 by orders of magnitude - see [this](https://github.com/cristianbuse/VBA-MemoryTools/issues/5#issuecomment-1384558244)). However, none of the 2 deal with releasing memory for objects or strings.

If you manage to replicate the problem then please replace ```MemCopy destPtr, srcPtr, size``` with ```CopyMemory ByVal destPtr, ByVal srcPtr, size```. If the problem is still there, then this is standard behaviour as ```CopyMemory``` is intended for working with unmanaged memory. However, if the problem goes away then please let me know and maybe raise an issue in my repo. Thanks!

**Comment by Theadd** [comment]
Hi @cristianbuse , that's an awesome elaborated post, thanks!

> use `GetArrayDims` on x64 only while on x32 will keep the original `GetArrayDimsCount`.

After seeing those results on x32, I totally agree. On the other hand, I wasn't even considering to include `GetArrayDimsNotNot` anywhere as publicly accessible, not because it's a bug, which I didn't know, but because it's not "generic" and you know I don't like to pollute the global scope 😉.

> please replace `MemCopy destPtr, srcPtr, size` with `CopyMemory ByVal destPtr, ByVal srcPtr, size`. If the problem is still there, then this is standard behaviour as `CopyMemory` is intended for working with unmanaged memory.

I indeed [managed to replicate the problem](https://github.com/Theadd/ds3x/blob/main/src/ds3x.accdb.src/modules/ArrayList.cls#L499C12-L527), reduced it to a single line of code which then I **downgraded it to be reproducible without any dependency**, other than `LibMemory`, ofc, which you can find below at the end of this post but, first, here is some ASCII depicted description of the steps taken to reproduce it.

```vb
'                       "Redim Preserve"        "Fill Memory with 0's to avoid releasing B"
'                       |                       |
' [A.B.C.D] -> [A.B.C.D.^] -> [A.B.B.C.D] -> [A._.B.C.D] -> [A.C.B.C.D]
'    ^                           ^ ^ ^                         ^
'    |                             ^ ^ ^                         "Copy back the memory range of the array "
'    |                              \ /                          "previously copied ByVal, also increasing"
'    ^"Insertion Index"              ^"Move Memory"              "refs count of safearray byref variants. "

aX.InsertRange InitialSize \ 4, aX.GetRange(InitialSize \ 2, InitialSize \ 4)
```

As far as I've digged in, there's a few conditions to meet in order to reproduce the crash using the depicted steps from above.
* There are some ByRef values being copied to/from two regions of memory.
* Those memory regions are contained one within the other or they're overlapping on one side.
* There's more than 4000-5000 elements in the `Variant()` array. (I've been slicing it by 1/4 of its size all the time tho.) 

Replacing the `MemCopy` call for its `CopyMemory` equivalent does not crash, neither using a separate allocated region do that memory movement of overlapping regions within a two-steps copy calls. Using `MemCopy` as the copy function for that two-steps copy method works as expected, no crash.

```diff
+RunWith InitialSize:=20000000, Mode:=MemAllocCopyMemoryMode
+RunWith InitialSize:=20000000, Mode:=MemAllocMemCopyMode
+RunWith InitialSize:=20000000, Mode:=CopyMemoryMode
+RunWith InitialSize:=2000, Mode:=MemCopyMode					' MemCopy + small array => OK
-RunWith InitialSize:=200000, Mode:=MemCopyMode					' MemCopy + large array => CRASH
```

Code used to reproduce: _(using current version of `ArrayList` in **main** branch)_
```vb
Public Sub RunWith(Optional ByVal InitialSize As Long = 200000, Optional ByVal Mode As MemoryMoveMode = MemoryMoveMode.MemCopyMode)
    Dim t() As Variant, i As Long
    ReDim t(0 To InitialSize - 1)
    For i = 0 To InitialSize - 1 Step 25: t(i) = CStr(i): Next i
    Dim aX As ArrayList: Set aX = ArrayListEx.Create(t).Instance
    MemoryMovingMode = Mode

    aX.InsertRange InitialSize \ 4, aX.GetRange(InitialSize \ 2, InitialSize \ 4)
End Sub
```
---

### Dependency-free version

Usage (with expected results):

```diff
+AtSafeArrayLevelVersion InitialSize:=20000000, UseMemCopy:=False 
+AtSafeArrayLevelVersion InitialSize:=2000, UseMemCopy:=True             ' MemCopy + small array => OK 
-AtSafeArrayLevelVersion InitialSize:=200000, UseMemCopy:=True           ' MemCopy + large array => CRASH 
```
Code to reproduce:
```vb
Private Declare PtrSafe Sub FillMemory Lib "kernel32" Alias "RtlFillMemory" (Destination As Any, ByVal Length As Long, ByVal Fill As Byte)

Public Sub AtSafeArrayLevelVersion(Optional ByVal InitialSize As Long = 200000, Optional ByVal UseMemCopy As Boolean = False)
    Static cSA As SAFEARRAY_1D
    
    Dim t() As Variant, i As Long
    ReDim t(0 To InitialSize - 1)
    For i = 0 To InitialSize - 1 Step 25: t(i) = CStr(i): Next i
    
    ' 2: Redim Preserve t (Moved step 2 before 1 just to reduce an additional step in case t gets reallocated)
    ReDim Preserve t(0 To UBound(t) + (InitialSize \ 4))
    
    ' 1: aX.GetRange(InitialSize \ 2, InitialSize \ 4)
    Dim cIndex As Long, cCount As Long, C() As Variant
    cIndex = InitialSize \ 2
    cCount = InitialSize \ 4
    With cSA
        .cDims = 1
        .cbElements = VARIANT_SIZE
        .fFeatures = FADF_VARIANT Or FADF_AUTO
        .cLocks = 10    ' Can't allow it to auto release as it's data is a memory range within `t`
                        ' (In ListRange (ArrayList's subrange), they're declared in a class module,
                        ' and it might be released before the container ArrayList, which will be
                        ' the one responsible of calling the final and only `Erase` on the data.
                        ' It has no use here as everythin is in the same scope, can be replaced by 0.
        .pvData = VarPtr(t(cIndex))
        .rgsabound0.cElements = cCount
    End With
    MemLongPtr(VarPtrArr(C)) = VarPtr(cSA)
    
    ' 3: Since C is overlapping t, assign it ByVal to C2
    Dim C2() As Variant
    C2 = C          ' Ref counts of elements within C get increased by 1
    
    Select Case UseMemCopy
        ' 4: The source of the crash, moving overlapping memory with elements ByRef using MemCopy
        Case True: MemCopy VarPtr(t(cIndex + cCount)), VarPtr(t(cIndex)), (InitialSize - cIndex) * VARIANT_SIZE
        ' 4b: The alternative CopyMemory sentence
        Case Else: CopyMemory ByVal VarPtr(t(cIndex + cCount)), ByVal VarPtr(t(cIndex)), (InitialSize - cIndex) * VARIANT_SIZE
    End Select
    
    ' 5: Zeroing the memory range where B2 will be copied to
    FillMemory ByVal VarPtr(t(cIndex)), cCount * VARIANT_SIZE, CByte(0)
    
    ' 6: Copy C2 back to where B initially was, ref counts of ByRef elements in C gets increased again by 1
    ' Replaced to reduce code: VariantArrayClone VarPtr(t(cIndex)), VarPtr(C2(0)), cCount
    ' Basic ByVal assignment loop alternative sentence
    For i = 0 To cCount - 1: t(cIndex + i) = C2(i): Next i
    
    ' Simulating `release` on `End Sub` of an inner procedure
    C2 = Array()    ' Ref counts of elements within C get decreased by 1
    With cSA
        .pvData = CLngPtr(0)        ' Prevent the safearray struct to release memory, which
        .rgsabound0.cElements = 0   ' wouldn't be needed here as it is static within a module
    End With
    
    ' ListRange's Class_Terminate (of GetRange() => C)
    MemLongPtr(VarPtrArr(C)) = CLngPtr(0)
    
    ' The container ArrayList Class_Terminate (aX)
    Erase t         ' Removes all elements within container ArrayList and deallocates t (releasing SafeArray and SafeArray Data)
End Sub
```

**Comment by cristianbuse** [comment]
Thanks @Theadd ,

I won't have time to test until Tuesday. However, I am thinking that the issue could be the buffer size.
I am curious if the crash still happens with a bigger buffer. This is the line:
[```Const chunkSize As Long = 2 ^ 16```](https://github.com/cristianbuse/VBA-MemoryTools/blob/1a2af4cf347365cd601c02c9efef4779c4daf663/src/LibMemory.bas#L723).
If you have time (before Tuesday) try setting the size to something like ```2 ^ 24``` and see if the crash still occurs. Thanks!

**Comment by Theadd** [comment]
Hi @cristianbuse , and thank you again.

> I am curious if the crash still happens with a bigger buffer.

You are right, as always, increasing that buffer avoids crash for bigger numbers, although, if I also increase the `InitialSize`, it crashes too.

I mean, with a chunkSize of `2 ^ 24`, an `InitialSize:=2000000` it won't crash but with an `InitialSize:=100000000` it will.

**Reply by cristianbuse** [reply]
Thank you @Theadd ,

This is very useful. It means that I incorectly assumed, when I wrote the method, that I can split the block arbitrarily. I will deal with this by either splitting differently (if even possible to figure a way) or default back to CopyMemory after the block size is too big. The split in chunks was purely for performance.

**Comment by cristianbuse** [comment]
Many thanks @Theadd ❤️ 

I fixed ```MemCopy``` with [this latest commit](https://github.com/cristianbuse/VBA-MemoryTools/commit/fbdd01146f510c5c640b6124149cc2b7df99abe4).

I took into account memory overlap but only for the entire block. I later introduced chunks when using SAFEARRAY (when BSTR is not possible i.e. would need to change the source) and I forgot to account for chunk overlap. Really happy you found the bug.

Anyway, I completely ditched the chunks as past a certain size it is faster to just call RtlMoveMemory directly. Basically anything up to 2 ^ 24 bytes gets copied in one go by BSTR/SAFEARRAY and anything larger uses the API directly.

**Comment by cristianbuse** [comment]
Hi @Theadd ,

Just thought that I should let you know that the way [the current CloneParamArray](https://github.com/cristianbuse/VBA-MemoryTools/blob/ac61dc988fed54164c3909b39b2ee2f42f1815ad/src/LibMemory.bas#L791-L844) was written was wrong.

The call simply fails if the first element in the ```ParamArray``` is a ByRef Variant because the entire memory copying then happens at the wrong address i.e. at the address of the actual Variant instead of the address of the first element of the ```ParamArray```. It just happened to work in my previous testing because I never had that first element as a Variant - until today.

Although I don't like the ```Not Not``` notation to get the ```SAFEARRAY``` address, it seems this is actually the only way to get the address for a param array because it cannot be coerced into a Variant. Basically, those 54 lines that I linked above can be re-written as:

```VBA
Public Sub CloneParamArray(ByVal paramPtr As LongPtr, ByRef out() As Variant)
    Static rm As REMOTE_MEMORY: If Not rm.isInitialized Then InitRemoteMemory rm
    RemoteAssign rm, paramPtr, rm.remoteVT, vbArray + vbVariant, out, rm.memValue
End Sub
```
while the call would look like this:
```VBA
Sub dh()
    Dim a As Variant: a = "ABC"
    Dim d As Variant: d = 4352436.75687
    TestCloneParam a, d
End Sub
Sub TestCloneParam(ParamArray args() As Variant)
    Dim arr() As Variant
    CloneParamArray Not Not args, arr '<--- the actual call
End Sub
```
I don't think you are using the ```CloneParamArray``` function but if you do then don't 😄 

### Update

I did eventually [push](https://github.com/cristianbuse/VBA-MemoryTools/commit/6a52b5b50796312dfc5d84e11e3b8c55fe4fd844) the above replacement.
